### PR TITLE
Add IterBasedRunner

### PR DESCRIPTION
**Pull request by hejm37**
This pull request adds iteration based runner `IterBasedRunner`.

### CODE REVIEW

File: mmcv/runner/hooks/lr_updater.py [hunk 1]
```diff
@@ -213,18 +240,79 @@ def get_lr(self, runner, base_lr):
...
+        cumulative_period (list[int]): Cumulative period list.
+
+    Returns:
+        int: The position of the right-closest number in the period list.
+    """
+    for i, period in enumerate(cumulative_period):
+        if iteration <= period:
+            return i
+
+
```
Comment (xvjiarui) [inline mmcv/runner/hooks/lr_updater.py:310]: This could be another PR.

File: mmcv/parallel/distributed.py [hunk 2]
```diff
@@ -8,3 +10,55 @@ class MMDistributedDataParallel(DistributedDataParallel):
...
+                self.reducer.prepare_for_backward(list(_find_tensors(output)))
+            else:
+                self.reducer.prepare_for_backward([])
+        else:
+            if torch.__version__ > '1.2':
+                self.require_forward_param_sync = False
+        return output
+
+    def val_step(self, *inputs, **kwargs):
+        # compatible with PyTorch 1.1 - 1.4
```
Comment (xvjiarui) [inline mmcv/parallel/distributed.py:41]: Similarly as above.

File: mmcv/runner/hooks/logger/text.py [hunk 3]
```diff
@@ -138,3 +145,49 @@ def log(self, runner):
 
         self._log_info(log_dict, runner)
         self._dump_log(log_dict, runner)
+
+
+@HOOKS.register_module
+class IterTextLoggerHook(TextLoggerHook):
+
+    def _log_info(self, log_dict, runner):
+        log_str = 'Exp Name: {}\t'.format(runner.meta['exp_name'])
```
Comment (xvjiarui) [inline mmcv/runner/hooks/logger/text.py:154]: Add check:
1. if meta is None
2. if 'exp_name' in runner.meta
3. if self.every_n_inner_iters(runner, self.interval_exp_name)

File: mmcv/runner/iter_based_runner.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            self._epoch += 1
+            if hasattr(self._dataloader.sampler, 'set_epoch'):
+                self._dataloader.sampler.set_epoch(self._epoch)
+            self.iter_loader = iter(self._dataloader)
+            data = next(self.iter_loader)
+
+        return data
+
+    def __len__(self):
+        return len(self._dataloader)
```
Comment (xvjiarui) [inline mmcv/runner/iter_based_runner.py:38]: What's the meaning or usage of `__len__` of `IterLoader`? 
It could be confusing since IterLoader will re-iter once hit the end.

File: tests/test_runner/test_runner.py [hunk 5]
```diff
@@ -38,22 +38,48 @@ def test_epoch_based_runner():
...
             pass
 
-        _ = EpochBasedRunner(model, batch_processor)
+        _ = EpochBasedRunner(
+            model, batch_processor, logger=logging.getLogger())
 
     with pytest.raises(TypeError):
         # batch_processor must be callable
-        model = OldStyleModel()
-        _ = EpochBasedRunner(model, batch_processor=0)
```
Comment (xvjiarui) [inline tests/test_runner/test_runner.py:46]: `OldStyleModel()` should be used. `batch_processor` shouldn't co-exist with `train_step`.

File: mmcv/runner/iter_based_runner.py (hunk 4 above)
Comment (hejm37) [reply mmcv/runner/iter_based_runner.py:38]: Since `torch.utils.data.DataLoader` has the `__len__` api, somewhere in the code that uses `IterLoader` to replace `torch.utils.data.DataLoader` would call this method. Thus, we expose this api.

File: mmcv/runner/hooks/logger/text.py [hunk 6]
```diff
@@ -138,3 +145,49 @@ def log(self, runner):
...
+        if runner.mode == 'train':
+            if isinstance(log_dict['lr'], dict):
+                lr_str = ''
+                for k, val in log_dict['lr'].items():
+                    lr_str += '{}: {:.4e} '.format('lr_' + k, val)
+                log_str += 'Epoch [{}][{}/{}]\t{}, '.format(
+                    log_dict['epoch'], log_dict['iter'], runner.max_iters,
```
[the remark this answers is not in the document]
Comment (hejm37) [reply mmcv/runner/hooks/logger/text.py]: I've merged the two loggers.

File: mmcv/runner/__init__.py [hunk 7]
```diff
@@ -16,14 +17,15 @@
...
-    'TextLoggerHook', 'TensorboardLoggerHook', 'WandbLoggerHook',
-    'MlflowLoggerHook', '_load_checkpoint', 'load_state_dict',
-    'load_checkpoint', 'weights_to_cpu', 'save_checkpoint', 'Priority',
-    'get_priority', 'get_host_info', 'get_time_str', 'obj_from_dict',
-    'init_dist', 'get_dist_info', 'master_only', 'OPTIMIZER_BUILDERS',
-    'OPTIMIZERS', 'DefaultOptimizerConstructor', 'build_optimizer',
+    'BaseRunner', 'Runner', 'EpochBasedRunner', 'IterBasedRunner', 'LogBuffer',
+    'HOOKS', 'Hook', 'CheckpointHook', 'ClosureHook', 'LrUpdaterHook',
+    'OptimizerHook', 'IterTimerHook', 'DistSamplerSeedHook', 'LoggerHook',
+    'PaviLoggerHook', 'TextLoggerHook', 'IterTextLoggerHook',
```
Comment (xvjiarui) [inline mmcv/runner/__init__.py:23]: CI failed due to `IterTextLoggerHook`.

Comment (hejm37) [reply mmcv/runner/__init__.py]: Fixed~
